### ISSUE TITLE
Release 2.28 checklist

### ISSUE THREAD

**Issue by rok-cesnovar** [stan-dev/cmdstan#1037]
Here is the checklist for the next release.

**Feature freeze (September 20th):**
- [ ] Create a release candidate tarball. Make sure RC has the new CMDSTAN_VERSION and that RC tarballs have binaries.
- [ ] Create Math/Stan RC releases. Make sure versions are updated.
- [ ] Create Stanc3 RC binary. Make sure versions are updated.
- [ ] Create a release candidate feature/bugfix list (this doesn't need to be polished, just so that RC testers know what to test)
- [ ] Run CmdstanR tests with the RC tarball (https://github.com/stan-dev/cmdstanr/runs/2628065416)
- [ ] Make a Discourse RC post
- [ ] Make a Twitter RC announcement

**Release (September 27th):**
- [ ] Create the Math Release notes
- [ ] Create the Stan Release notes
- [ ] Create the Cmdstan Release notes
- [ ] Create the Stanc3 Release notes
- [ ] Create the Math release
- [ ] Create the Stan release
- [ ] Create the Stanc3 release
- [ ] Create the CmdStan binaries (make sure the version is incremented and tarball includes binaries)
- [ ] Run CmdStanR tests with the release tarball
- [ ] Make a Discourse release announcement post
- [ ] Cross-post the Discourse announcement on the Stan blog (dont forget thanking our sponsors)
- [ ] Make the Twitter announcement
- [ ] Update Math/Stan/CmdStan versions on develop

**Comment by rok-cesnovar** [comment]
Tagging @bob-carpenter @SteveBronder @WardBrian @nhuurre @spinkney that have been doing stuff across the Math/Stan/Cmdstan/Stanc3 repositories in the last few weeks. Does any one of you have anything they want to get in for the next release and need more eyes on it or some help for a two week push? The deadlines are generally hard but we can be flexible if there is months-long project that might miss for a day or two.

I currently do not have any open active issues/project that I would like to finish for this release.

I will mostly handle the tasks listed above with @serban-nicusor-toptal and will tag devs to double check when the time comes.

**Comment by WardBrian** [comment]
I have two small stanc3 PRs I would like included. The first is https://github.com/stan-dev/stanc3/pull/959, which adds the ability to write complex numbers as `3 + 2i`, and the second is https://github.com/stan-dev/stanc3/pull/952, which just changes how we escape string literals in our generated code. 

Both have docs that have been merged and are passing all tests, so all they need is review. They're pretty small (3 files/~50loc each)

Nothing else from me!

**Comment by spinkney** [comment]
- generalized inverse fix for symmetric matrices is done. Needs review https://github.com/stan-dev/math/pull/2577
- the multi normal derivatives just calls multi_normal_cholesky under the hood it is done except that the expression tests are failing due to different expressions for multi_normal vs multi_normal_cholesky. I need someone to help with that. https://github.com/stan-dev/math/pull/2572

**Comment by wds15** [comment]
@SteveBronder … varmat???

**Comment by SteveBronder** [comment]
@wds15 I think the plan is to release it immediately after 2.28 so we have a full dev cycle for testing etc.

**Comment by rok-cesnovar** [comment]
I have a few minor cleanup PRs ready if anyone has time to review:
https://github.com/stan-dev/cmdstan/pull/1039
https://github.com/stan-dev/cmdstan/pull/1040

**Comment by serban-nicusor-toptal** [comment]
Hey @rok-cesnovar should I wait until tomorrow for #1040 to be merged in ?

**Comment by rok-cesnovar** [comment]
Will fix that forst thing in the morning and then we can proceed yeah. The deadline is Tuesdya morning CET time anyways.

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal we need to postpone until we address https://github.com/stan-dev/stan/issues/3057 so you are free today :)

**Comment by SteveBronder** [comment]
@rok-cesnovar should we also wait till we support Eigen 3.4? I'm working on that right now and there's a weird bug with `mdivide_left/right` I'm trying to sort out. Seems like it would be good to make sure we are compatible so that when R etc. bumps up everything works smoothly

**Comment by SteveBronder** [comment]
Oh I'd also like to update the OpenCL hpp headers so they work with OpenCL 3.0 devices. Opened up a PR for that below

https://github.com/stan-dev/math/pull/2587

**Comment by wds15** [comment]
R is not really an urgent consideration given how much rstan is behind is what I would say. So we do not need eigen 3.4 soon.

**Comment by rok-cesnovar** [comment]
> should we also wait till we support Eigen 3.4?

Just doing it for R is not needed. As Sebastian notes, there is almost zero chance rstan will get to 2.28 until the next release. If there are other bigger reasons or gains we get with that for 2.28 we can do it.

> Oh I'd also like to update the OpenCL hpp headers so they work with OpenCL 3.0 devices. Opened up a PR for that below

Will review, seems tests are failing with it.

**Comment by rok-cesnovar** [comment]
The main things I will highlight in the release candidate as things we wish "power" users would test and give feedback on and are user-facing:

- complex numbers
- updated Power/Newton algebra solvers & faster `inv_Phi`
- automatic handling of precompiled headers on Windows with RTools 4.0
- improved error messaging/type error explanations in stanc3
- improve pretty printing/auto formatting

There really isnt a ton of things this time around, at least not stuff that is user-facing atm.

@bob-carpenter @WardBrian do you have any sort of example, model + data for the complex stuff? No need to explain the model or have any accompanying text. Just something "power" users that typically help us test the release candidate can run and check that it works for them.

For the rest I think I have enough knowledge of what was done so I can write a sentence on it.

**Comment by rok-cesnovar** [comment]
The fix for https://github.com/stan-dev/stan/issues/3057 is in so once that propagates upstream to cmdstan  (10 hours or so) we are good to go in terms of the release candidate. But we dont want to make a RC on the weekend but we can do it first thing on Monday.

**Comment by wds15** [comment]
And threaded sampling of multiple chains is in cmdstan now…that’s a nice user feature.

**Comment by rok-cesnovar** [comment]
Oh right right, how can I forget that. Thanks! Added to the list above.

**Comment by SteveBronder** [comment]
@bgoodri @hsbadr during the feature freeze should we run the rstan reverse dependencies? Think we could make things a bit easier if we do any patches for rstan during the freeze

**Comment by hsbadr** [comment]
> during the feature freeze should we run the rstan reverse dependencies? Think we could make things a bit easier if we do any patches for rstan during the freeze

As far as I'm concerned, the experimental version of RStan is ok while CRAN issues can't be addressed with this release, considering the big version gap and backward incompatibility at the C++ level. Though, it may be interesting to run reverse dependency checks based on the experimental version of RStan as a way to extensively test Stan v2.8 RC.

**Comment by wds15** [comment]
Are we tagging today the RC?

**Comment by rok-cesnovar** [comment]
Hopefully yes if @serban-nicusor-toptal has the time today. I have the texts mostly ready, apart from the complex number examples.

**Comment by WardBrian** [comment]
Bob put a complex regression model as an example in the complex numbers chapter, but I don't think we have data for it

**Comment by rok-cesnovar** [comment]
Oh I missed that. Thanks. I guess it will do without the data as well.

**Comment by serban-nicusor-toptal** [comment]
Find the release candidates below:
- [stanc3 2.28.0-rc1](https://github.com/stan-dev/stanc3/releases/tag/v2.28.0-rc1)
- [math 4.2.0-rc1](https://github.com/stan-dev/math/releases/tag/v4.2.0-rc1)
- [stan 2.28.0-rc1](https://github.com/stan-dev/stan/releases/tag/v2.28.0-rc1)
- [cmdstan 2.28.0-rc1](https://github.com/stan-dev/cmdstan/releases/tag/v2.28.0-rc1) ( currently waiting on [stanc3 binaries build](https://jenkins.mc-stan.org/blue/organizations/jenkins/stanc3/detail/v2.28.0-rc1/1/pipeline) for the tarball )

As soon as the stanc3 binaries are done and cmdstan tarball uploaded, I will edit my post.

**Comment by serban-nicusor-toptal** [comment]
Hey @rok-cesnovar everything should be ready now, can I please ask you to double check when you have a minute? Thank you!

**Comment by hsbadr** [comment]
> ```
> Error: Error in loadNamespace(x) : there is no package called ‘rcmdcheck’
> ```

It failed to install `rcmdcheck` because the `configure` script can't find `libcurl`.

```r
------------------------- ANTICONF ERROR ---------------------------
Configuration failed because libcurl was not found. Try installing:
 * deb: libcurl4-openssl-dev (Debian, Ubuntu, etc)
 * rpm: libcurl-devel (Fedora, CentOS, RHEL)
 * csw: libcurl_dev (Solaris)
If libcurl is already installed, check that 'pkg-config' is in your
PATH and PKG_CONFIG_PATH contains a libcurl.pc file. If pkg-config
is unavailable you can set INCLUDE_DIR and LIB_DIR manually via:
R CMD INSTALL --configure-vars='INCLUDE_DIR=... LIB_DIR=...'
--------------------------------------------------------------------
ERROR: configuration failed for package ‘curl’
```

@rok-cesnovar Can you install `libcurl-devel` in the GHA workflow for Ubuntu?

**Comment by rok-cesnovar** [comment]
Hey,

in general looks great, there are a few things to cleanup:
- we will only supply the linux-stanc, windows-stanc and mac-stanc files in the cmdstan-2.28.0-rc1.tar.gz tarball, the rest of them we will build separate tarballs, but that isnt urgent
- there are files ._windows-stanc, ._mac-stanc, etc in the bin folder
- there is a ._cmdstan file in the top folder whjen I unarchive

@hsbadr will do, thanks!

**Comment by serban-nicusor-toptal** [comment]
@rok-cesnovar fixing that now

**Comment by WardBrian** [comment]
We will be producing separate cmdstan tarballs for the different linux archs, correct?

**Comment by rok-cesnovar** [comment]
Yeah, similar to how we have them already for arm: https://github.com/stan-dev/cmdstan/releases/tag/v2.27.0

**Comment by serban-nicusor-toptal** [comment]
@rok-cesnovar I've removed not needed binaries and excluded `._*` files and `.DS_Store` ones, they're from `tar` and can be ignored by `--disable-copyfile`, example:
`tar --disable-copyfile --exclude='.DS_Store' -czvf cmdstan-2.28.0-rc1.tar.gz cmdstan/`

**Comment by wds15** [comment]
@rok-cesnovar should we announce the RC on the forum? FYI, I just ran my usual model over it and it looks as if things are good this time.

**Comment by rok-cesnovar** [comment]
Yeah, I am in the process of doing that. It got a bit late yesterday but it looks good on my end as well mostly and I am just finishing up the texts. Will post on forum/twitter.

**Comment by rok-cesnovar** [comment]
An issue popped up that got caught by some upstream tests in cmdstanr https://github.com/stan-dev/cmdstan/issues/1043

I think it makes sense to make and RC2 and then post.

**Comment by serban-nicusor-toptal** [comment]
If there are no outstanding issues I will start the release for RC2 in 10 minutes.

**Comment by rok-cesnovar** [comment]
No no, wait, we need to merge the fix first. Will tag you here.

**Comment by wds15** [comment]
I don‘t think we need another rc just for that…thats not so severe from my view.

**Comment by rok-cesnovar** [comment]
Yeah, not sever, but kind of cant run any threaded model without modifying your scripts, which is a bit annoying. And we only need to rebuild the tarball so should be simple.

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal go ahead and build another RC tarball

**Comment by serban-nicusor-toptal** [comment]
@rok-cesnovar which one do we need now?:
- create cmdstan and stanc3 rc2 then build tarball
- create stanc3 rc2 and re-do the cmdstan rc1 tarball 
I'm asking to make sure I understood correctly, thanks!

**Comment by rok-cesnovar** [comment]
No need to rebuild stanc3. Just cmdstan rc2 and rebuild tarball.

**Comment by serban-nicusor-toptal** [comment]
Thank you! I'm on it

**Comment by serban-nicusor-toptal** [comment]
cmdstan [v2.28.0-rc2](https://github.com/stan-dev/cmdstan/releases/tag/v2.28.0-rc2) released and tarball rebuild with new source code and stanc3 _v2.28.0-rc1_ binaries.

**Comment by rok-cesnovar** [comment]
Thanks, will just re-run a few tests again and then proceed with posting an annoucement.

**Comment by rok-cesnovar** [comment]
The annoucements are now posted. Sorry everyone for a bit of a delay. Lets hope the RC period goes by smoothly. 

Thanks to all that helped out!

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal can you make a new RC for stanc3 with the current master. We need it to test the fixes that had to be done for pystan3. If you have the time also make RC3 for cmdstan. Thanks!

**Comment by serban-nicusor-toptal** [comment]
stanc3 [v2.28.0-rc2](https://github.com/stan-dev/stanc3/releases/tag/v2.28.0-rc2) is currently [building binaries](https://jenkins.mc-stan.org/blue/organizations/jenkins/stanc3/detail/v2.28.0-rc2/1/pipeline), right after that is done I will do rc for cmdstan with the binaries. I will reply here as soon as it's done

**Comment by wds15** [comment]
If this upgrade does not delay the release, then that's all great... I am eagerly waiting for the final thing to put this 2.28 into our about to be released production system... so I myself would appreciate no delays...just to make my personal motivations clear, which are not meant to make up a decision for this.

**Comment by rok-cesnovar** [comment]
It should not effect the release so I would say go for it Nic.

**Comment by serban-nicusor-toptal** [comment]
cmdstan [v2.28.0-rc3](https://github.com/stan-dev/cmdstan/releases/tag/v2.28.0-rc3) is now out with the tarball containing stanc3 [v2.28.0-rc2](https://github.com/stan-dev/stanc3/releases/tag/v2.28.0-rc2) binaries.
Related to the Jenkins upgrade, it's very low chance but things might happen with a plugin upgrade or some incompatibility. Since there's not a big rush with it, I will postpone it for next weekend (9-10 Oct) just so we can avoid any unexpected delivery issues Monday and have time to properly finish the release. If everyone is okay with this I will post a thread on the forums sometimes today ( developers section ) for 9-10 Oct weekend.

**Comment by rok-cesnovar** [comment]
The release notes for Math/Stan/Cmdstan are available here: https://gist.github.com/rok-cesnovar/714c5620b3f5d99bb57920e66a940acb

Would appreciate if someone gave that a quick look.

They were generated by https://github.com/stan-dev/math/blob/develop/scripts/release_notes.py and then just grouped and edited a bit.

**Comment by rok-cesnovar** [comment]
Stanc3 release notes are in the same gist as a comment: https://gist.github.com/rok-cesnovar/714c5620b3f5d99bb57920e66a940acb#gistcomment-3915250

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal if there will be no urgent bug reports until the end of today, please start with the release tomorrow morning or anytime after when it fits your schedule. 

The release notes for the 4 repositories are linked in the above 2 comments. I think all docs pull requests that were meant for this release cycle were merged, so you can rebuild that as well.

**Comment by serban-nicusor-toptal** [comment]
Will do @rok-cesnovar, gonna try to do it before US wakes up. Thanks for having everything ready!
The release notes look good to me, one aspect (not important) question:
![image](https://user-images.githubusercontent.com/48496524/135829687-668ffa0c-5a31-4f53-84be-52a28aaaf441.png)
Shouldn't the last two be under the first OpenCL column ?

**Comment by rok-cesnovar** [comment]
Thanks Nic, fixed.

**Comment by serban-nicusor-toptal** [comment]
If there are no outstanding bugs/prs I will start the release process in 15-20 minutes.

**Comment by rok-cesnovar** [comment]
Go ahead and start yes. Thanks!

**Comment by serban-nicusor-toptal** [comment]
Most of the work is done, I am now waiting for [stanc3 binaries](https://jenkins.mc-stan.org/blue/organizations/jenkins/stanc3/detail/v2.28.0/1/pipeline), right after that I'll make the tarball and "publish" the latest release in github.

**Comment by rok-cesnovar** [comment]
Great, the CmdStan tarball seems to be working, I started a upstream tests with it to double check. Once we update the docs for 2.28 I think your part of the release is finished. 

Once that is done I will finish up the release text with the new links to the actual documentation and post it on the blog/discourse.

**Comment by WardBrian** [comment]
I’m not yet seeing the non-x86 tarballs for cmdstan, are those done later/manually?

**Comment by serban-nicusor-toptal** [comment]
Everything should be done except two things:
- Docs, there seems to be a package mismatch in the pipeline, I'm on it.
- non-x86 tarball for cmdstan, seems I totally forgot about it. If you guys can give me a quick how to I'll make it right away. Thanks!

**Comment by rok-cesnovar** [comment]
Previously we only did the -arm64 one and I built that one locally once the tests confirmed the x86 is all green lights. 

I can do it as well this time for the 6 architectures we support and will post the script so we can automate this for 2.29 @serban-nicusor-toptal if that is fine with you. And you can focus on fixing the docs.

**Comment by WardBrian** [comment]
Guessing docs issue is related to https://github.com/stan-dev/docs/pull/390

**Comment by serban-nicusor-toptal** [comment]
That's okay with me @rok-cesnovar thanks a lot!
Yes @WardBrian thanks for the link as it has some useful information.

**Comment by rok-cesnovar** [comment]
> seems I totally forgot about it. 

Slipped my mind as well. I updated the checklist in the top post to explicitly state we need x86 and non-x86 tarballs so we dont forget next time. Will update the template in stan-dev/ci-scripts once the release is over.

**Comment by serban-nicusor-toptal** [comment]
I'm on it, just takes a bit to compile and install all the stuff.
FYI @WardBrian our (Jenkins) uses mostly stateful machines that we manage ( linux, mac and windows (down atm)) and in AWS we have some on-demand Linux and Windows (no gpu) that use a templated AMI (launch template).
TL;DR; I need to login into the machine itself, make it work there and after the CI should work just fine.

**Comment by serban-nicusor-toptal** [comment]
We're good with docs now.
Pipeline can be found [here](https://jenkins.mc-stan.org/blue/organizations/jenkins/BuildDocs/detail/BuildDocs/80/pipeline/32), the last step failed because `make manual` is not part of the makefile anymore.

If someone has the time, please review and approve the MRs:
- [stan-dev.github.io](https://github.com/stan-dev/stan-dev.github.io/pull/160)
- [docs](https://github.com/stan-dev/docs/pull/416)

---

I've ran the R command manually and figured out what was missing/wrong version:
- R Package bookdown must be version 2.23, see README for build tools requirements
- Please install or upgrade Pandoc to at least version 1.17.2; or if you are using RStudio, you can just install RStudio 1.0+.
packageVersion("snow")
- R install.packages("kableExtra")
- R install.packages("dplyr")

Please let me know if there's anything else to be done! Thanks for all the help with this release !!!

**Comment by serban-nicusor-toptal** [comment]
I see now that after pipelines are done we have a few failed tests:
- 2 in math [OpenCL GPU tests](https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/master/80/pipeline/152)
- 1 in stan [Mac unit tests](https://jenkins.mc-stan.org/blue/organizations/jenkins/Stan/detail/master/27/pipeline) and 1 on [Github (Ubuntu)](https://github.com/stan-dev/stan/runs/3803908911)

**Comment by rok-cesnovar** [comment]
> 2 in math OpenCL GPU tests

These are test machine config errors (it cant find the GPU), restarted it and it will hopefully pickup a different machine. Will check later.

> 1 in stan Mac unit tests and Github (Ubuntu)

These are failing tests for version. You forgot to change the version in the test: https://github.com/stan-dev/stan/blob/master/src/test/unit/version_test.cpp I am not really sure this test is needed, but that is a different story :)

**Comment by serban-nicusor-toptal** [comment]
Argh I'm not sure how that escaped, should I do a hotfix pr in master or just develop for now ?

**Comment by rok-cesnovar** [comment]
I think it makes sense to hotfix that yeah.

**Comment by serban-nicusor-toptal** [comment]
I've created a pr to merge master into develop [first](https://github.com/stan-dev/stan/pull/3069) and after that one a [second](https://github.com/stan-dev/stan/pull/3068) pr which will apply just the hotfix, after that to keep it clean I will release with the new commit hash.

**Comment by rok-cesnovar** [comment]
Thanks! Approved both, they look good. Will make a note in the checklist that we should run these pre-release so its less stressful.

**Comment by rok-cesnovar** [comment]
> I’m not yet seeing the non-x86 tarballs for cmdstan, are those done later/manually?

All six of the non-x86 tarballs are now uploaded. So I think with that all that is left to do is make the announcements.

**Comment by serban-nicusor-toptal** [comment]
Just one moment so I can update stan in cmdstan submodules and then create for the rest master-> develop. ty

**Comment by rok-cesnovar** [comment]
I am not going to do the annoucements today as its getting a bit late here for me to work on release texts :) will post them in the morning so no rush.

**Comment by serban-nicusor-toptal** [comment]
Everything should be fine now, there's only [this PR](https://github.com/stan-dev/math/pull/2598) left which will update develop with code from master for Math. 
Sorry for my unintended typo earlier.

**Comment by rok-cesnovar** [comment]
Its all good Nic. Thanks for all your hard work!

**Comment by SteveBronder** [comment]
Yes much appreciated Nic!

**Comment by wds15** [comment]
Great. For the notes: It is Math 4.2.1.

**Comment by rok-cesnovar** [comment]
The Math library tests have now finally passed and the develop state of Math can be released as well Stan. CmdStan will be updated in about 3 hours.

**Comment by serban-nicusor-toptal** [comment]
Nice! If there's nothing else tomorrow morning I'll do the patch release.

**Comment by serban-nicusor-toptal** [comment]
Hey @rok-cesnovar everything should be in place.

- [stanc3 v2.28.1](https://github.com/stan-dev/stanc3/releases/tag/v2.28.1)
- [Math v4.2.1](https://github.com/stan-dev/math/releases/tag/v4.2.1)
- [Stan v2.28.1](https://github.com/stan-dev/stan/releases/tag/v2.28.1)
- [CmdStan v2.28.1](https://github.com/stan-dev/cmdstan/releases/tag/v2.28.1)

When you have a few minutes, please review `master->develop`:
- [Math](https://github.com/stan-dev/math/pull/2606)
- [Stan](https://github.com/stan-dev/stan/pull/3072)
- [CmdStan](https://github.com/stan-dev/cmdstan/pull/1052)

**Comment by rok-cesnovar** [comment]
The annoucements are out as well https://twitter.com/mcmc_stan/status/1451200757219291150

I also built and uploaded non-x86 tarballs.
